### PR TITLE
Add AddVisitor method and Configure hook

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Fixes https://github.com/microsoft/typespec/issues/3827

Also adds a Configure hook so that AddVisitor can be called directly from the Plugin. Considered just using the ctor but that would mean we would have virtual property calls in a ctor.

### CODE REVIEW

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/OutputLibraryTests.cs [hunk 1]
```diff
@@ -26,9 +28,54 @@ public void BuildTypeProviders_Override()
...
+        public void CanOverrideGetOutputLibraryVisitors()
+        {
+            var outputLibrary = new MockOutputLibraryOverridingVisitors(new [] { new MockOutputLibraryVisitor() });
+            Assert.AreEqual(1, outputLibrary.GetOutputLibraryVisitors().Count());
+
+            outputLibrary.AddVisitor(new MockOutputLibraryVisitor());
+            Assert.AreEqual(2, outputLibrary.GetOutputLibraryVisitors().Count());
+        }
+
+        private class MockOutputLibraryVisitor : OutputLibraryVisitor
```
Comment (m-nash) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/OutputLibraryTests.cs:48]: nit: can we call this TestOutputLibraryVisitor so its not confused with a Moq visitor.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/OutputLibraryTests.cs [hunk 2]
```diff
@@ -26,9 +28,54 @@ public void BuildTypeProviders_Override()
...
+        private class MockOutputLibrary : OutputLibrary
         {
-            public MockOutputLibrary() : base() { }
+            protected override TypeProvider[] BuildTypeProviders()
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        private class MockOutputLibraryOverridingVisitors : OutputLibrary
```
Comment (m-nash) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/OutputLibraryTests.cs:60]: same here with Mock vs Test, or actually make this a mock by overloading mockhelpers with new lamda params

File: packages/http-client-csharp/generator/SamplePlugin/src/SampleCodeModelPlugin.cs [hunk 3]
```diff
@@ -14,6 +14,9 @@ public class SampleCodeModelPlugin(GeneratorContext context) : ClientModelPlugin
     {
         public override SamplePluginTypeFactory TypeFactory { get; } = new SamplePluginTypeFactory();
 
-        public override OutputLibrary OutputLibrary { get; } = new SamplePluginOutputLibrary();
+        public override void Configure()
+        {
+            OutputLibrary.AddVisitor(new SamplePluginOutputLibraryVisitor());
```
Comment (m-nash) [inline packages/http-client-csharp/generator/SamplePlugin/src/SampleCodeModelPlugin.cs:19]: I almost feel like Configure should take in the visitors and do this internally?

**Review by m-nash — APPROVED** [review]

File: packages/http-client-csharp/generator/SamplePlugin/src/SampleCodeModelPlugin.cs (hunk 3 above)
Comment (JoshLove-msft) [reply packages/http-client-csharp/generator/SamplePlugin/src/SampleCodeModelPlugin.cs:19]: This is one hook that lets customers add the visitor. They can also derive from OutputLibrary and override GetVisitors.

**Pull request merged**